ocaml/xapi/xapi_xenops.ml

```diff
@@ -647,29 +647,30 @@ module MD = struct
     let pci_address = Db.PCI.get_pci_id ~__context ~self:pci in
     Xenops_interface.Pci.address_of_string pci_address
 
+  let get_virtual_pci_address ~__context vgpu =
+   let device = vgpu.Db_actions.vGPU_device in
+   let virtual_pci_address  =  Xapi_globs.vgpu_pci_prefix ^ device ^".0" in
+   Xenops_interface.Pci.address_of_string virtual_pci_address
+
   let of_nvidia_vgpu ~__context vm vgpu =
     let open Vgpu in
     (* Get the PCI address. *)
     let physical_pci_address = get_target_pci_address ~__context vgpu in
-    (* Get the vGPU config. *)
+    let virtual_pci_address = get_virtual_pci_address ~__context vgpu in
+    (* Get the vGPU type_id, since psubdev_id is option, can't decide the string format *)
     let vgpu_type = vgpu.Db_actions.vGPU_type in
-    let internal_config =
-      Db.VGPU_type.get_internal_config ~__context ~self:vgpu_type in
-    let config_file =
-      try List.assoc Xapi_globs.vgpu_config_key internal_config
-      with Not_found -> failwith "NVIDIA vGPU config file not specified"
-    in
-    let config_file =
+    let identifier = Db.VGPU_type.get_identifier ~__context ~self:vgpu_type in
+    let type_id =
```

Comment:
The identifier is not meant to be parsed. Add the type_id to the `internal_config` field of `VGPU_type`, as was done for the conf file.

Reply:
Yes, I will remove from `identifier` to `internal_config.` Thanks.